base/compiler/abstractinterpretation.jl

```diff
@@ -1232,6 +1232,36 @@ function abstract_call_builtin(interp::AbstractInterpreter, f::Builtin, (; fargs
                 end
                 return Conditional(aty.var, ifty, elty)
             end
+        elseif f === isdefined
+            uty = argtypes[2]
+            a = ssa_def_slot(fargs[2], sv)
+            if isa(uty, Union) && isa(a, SlotNumber)
+                fld = argtypes[3]
+                vtype = Union{}
+                elsetype = Union{}
+                if isa(fld, Const)
+                    fld = fld.val
+                end
+                if !isa(fld, Symbol)
+                    # even when isdefined is invalid, can filter out types with no fields
+                    for ty in uniontypes(uty)
+                        length(fieldnames(ty)) == 0 && continue
+                        vtype = tmerge(vtype, ty)
+                        elsetype = tmerge(elsetype, ty)
+                    end
+                    return Conditional(a, vtype, elsetype)
+                else
+                    for ty in uniontypes(uty)
+                        cnd = hasfield(ty, fld)
+                        if cnd
+                            vtype = tmerge(vtype, ty)
+                        else
+                            elsetype = tmerge(elsetype, ty)
+                        end
+                    end
+                    return Conditional(a, vtype, elsetype)
+                end
```

Comment:
```suggestion
                for ty in uniontypes(uty)
                    cnd = isdefined_tfunc(ty, fld)
                    if isa(cnd, Const)
                        if cnd.val::Bool
                            vtype = tmerge(vtype, ty)
                        else
                            elsetype = tmerge(elsetype, ty)
                        end
                    else
                        vtype = tmerge(vtype, ty)
                        elsetype = tmerge(elsetype, ty)
                    end
                end
                return Conditional(a, vtype, elsetype)
```
I still think something like this would be cleaner ?
Also, `isdefined` can take `::Int` argument, so we may want to handle that case too.
`isdefined_tfunc` already implements such cases as well, so I'd suggest using it here.

Reply:
Hmm, you're totally right. Not sure why the previous version with `isdefined_tfunc` wasn't working locally for me. This passes all tests and is certainly more versatile, so going with this approach.